### ISSUE TITLE
Missing lines requested and send multiple times on fast scrolling

### ISSUE THREAD

**Issue by rkusa** [google/xi-mac#34]
Example console output:

```
requesting missing: 6232..6297
requesting missing: 6598..6663
requesting missing: 6767..6832
requesting missing: 7222..7287
requesting missing: 7489..7554
requesting missing: 7612..7677
requesting missing: 7939..8004
requesting missing: 8033..8098
requesting missing: 8121..8186
requesting missing: 8279..8344
requesting missing: 8414..8479
requesting missing: 8473..8538
requesting missing: 8473..8538
requesting missing: 8473..8538
requesting missing: 8473..8538
requesting missing: 8473..8538
requesting missing: 8473..8538
requesting missing: 8473..8538
requesting missing: 8473..8538
requesting missing: 8473..8538
requesting missing: 8473..8538
requesting missing: 8473..8538
requesting missing: 8473..8538
requesting missing: 8473..8538
requesting missing: 8473..8538
requesting missing: 8473..8538
requesting missing: 8473..8538
requesting missing: 8473..8538
requesting missing: 8473..8538
requesting missing: 8473..8538
requesting missing: 8473..8538
requesting missing: 8473..8538
requesting missing: 8473..8538
requesting missing: 8473..8538
requesting missing: 8473..8538
requesting missing: 8473..8538
requesting missing: 8480..8538
requesting missing: 8480..8538
```

Since a request of missing lines is send on each draw call, requests for the same or overlapping ranges are sometimes send multiple times, because the response for the previous request has not arrived yet.

My initial expectation was, that even though the `core` receives requests for the same range multiple times, it only sends the lines once (due to its cache of `valid_lines`). However, subsequent `request_lines` requests to the same range seem to invalidate the range and, thus, the lines are resend.

Possible solutions are: Send scroll area and have the core push the lines instead of requesting lines, or make the FE to keep track of in-flight requests.

**Comment by raphlinus** [comment]
This optimization should be done by the FE, not the core. The core has no way of knowing whether the re-request is redundant, or it is because the FE has evicted the lines from its cache. The FE, on the other hand, can distinguish these cases.

Cut'n'paste from IRC:

<raph> having given it a bit of thought, i think the single most important thing that can improve scrolling performance is for the FE to not constantly re-request the same line when it's invalid
1:33 PM conceptually, the way to do this is for a line to have three states: valid, invalid, and requested
1:33 PM (in the FE's line cache)
1:34 PM when trying to draw, send a request for all lines in invalid (but not requested) state
1:34 PM E<eyelash-M> why does the fe even have to request lines?
1:35 PM can't it just send the scroll and the core will send an update?
1:35 PM R<•raph> Raph Levien yes, that too
1:35 PM E<eyelash-M> (that's what xi-gtk does anyway)
1:36 PM R<•raph> Raph Levien this was quite different in the macOS 10.11 days
1:36 PM there, Cocoa would allocate large tiles outside the actually visible scroll region
1:36 PM and ask the app to draw into them
1:37 PM this is the main reason why it has to deal so much with lines not in the scroll region
1:38 PM but this has changed in 10.12; one consequence is that cocoa asks the app to draw a lot more often, at a much finer grain
1:38 PM (so, to do this in a fully general way, when the FE sends a scroll notification, it can/should also put all invalid lines in the scroll region into the "requested" state)
1:39 PM this means that the FE will never request lines that are in the scroll region, but if the framework asks for lines outside the scroll region, they will eventually get drawn correctly

**Comment by cmyr** [comment]
This was fixed in #89.
